### PR TITLE
Zp/m1w5 scope

### PR DESCRIPTION
**Pull request by zoefarrell**

### CODE REVIEW

File: Mod1/Lessons/Week5/Scope.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
+# Scope and Access Modifiers
+
+## Learning Goals
+* Identify how **scope** affects where variables can be accessed
+* Have a reference to look at when **scope** errors show up
+* Understand the difference between the **public** and **private** access modifiers
```
Comment (zoefarrell) [inline Mod1/Lessons/Week5/Scope.md:6]: I realized while working on this that when to use public vs private is actually quite a nuanced and advanced topic. Because it works to leave everything public, it's just better to make things that should be private, private. 

I think it's good to start thinking about how others interact with your code, not sure if by week 5 students will be ready? I'm thinking I'll split this into two lessons and we can drop them into the schedule when we think students are ready for them?

File: Mod1/Lessons/Week5/Scope.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+## Practice
+
+Here is a modified version of some code from our Class Interaction Lesson. 
+
+>Take 3 minutes: For each variable defined, determine the box for the scope where that variable is available.
+
+<details><summary>Solution</summary><br/>
+
+![Scope Example 3 With Highlighting](/Mod1/Images/Week5/ScopeExample3.png)
```
Comment (zoefarrell) [inline Mod1/Lessons/Week5/Scope.md:59]: I realized while working on this that it might not be readable to folks that are colorblind. It's already a lot going on for anyone to look at. Let me know if you have suggestions for readability here!

File: Mod1/Lessons/Week5/Scope.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+## Learning Goals
+* Identify how **scope** affects where variables can be accessed
+* Have a reference to look at when **scope** errors show up
+* Understand the difference between the **public** and **private** access modifiers
+
+# Scope
+
+## Intro
+Scope is what you have access to and where you have access to it.
```
Comment (zoefarrell) [inline Mod1/Lessons/Week5/Scope.md:11]: I pulled this intro from the turing backend scope lesson, but looking back on it now I wonder if it's too wordy. I think in general I write too many words, let me know if you have suggestions for things to cut!

File: Mod1/Lessons/Week5/Scope.md (hunk 1 above)
Comment (memcmahon) [reply Mod1/Lessons/Week5/Scope.md:6]: I agree - this can be very nuanced.  To me, there are two 'learning goals' around public/private:
  * definition: relatively simple, Mod1 students should be ready to master this.
  * use cases: very nuanced, something that could be another lesson in a later mod, or something that students build up through practice and on the job.

File: Mod1/Lessons/Week5/Scope.md (hunk 3 above)
Comment (memcmahon) [reply Mod1/Lessons/Week5/Scope.md:11]: I don't think this is too wordy!  But, I do think we might want to get students mental juices flowing before we dive right in.  Maybe some kind of warm-up exercise that gets them thinking about when _they_ have experienced scope issues.  It could be as simple as 'here's a screenshot of a time when I didn't have access to the variable I wanted'... have you encountered this before? What did you do to fix the issue?

File: Mod1/Lessons/Week5/Scope.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+Note that a variable is accessible **everywhere** inside the code block in which it's defined, even from within a nested code block.
+
+## Practice
+
+Here is a modified version of some code from our Class Interaction Lesson. 
+
+>Take 3 minutes: For each variable defined, determine the box for the scope where that variable is available.
+
+<details><summary>Solution</summary><br/>
```
Comment (memcmahon) [inline Mod1/Lessons/Week5/Scope.md:57]: 🤯 

Love this 'hiding' of the solution.  To me, it might make more sense to put it below the stevie code.

File: Mod1/Lessons/Week5/Scope.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+private string Role;
+```
+But when would you use `public` vs `private`? Why are access modifiers so helpful? Let's see an example!
+
+## Why are access modifiers helpful?
+
+Imagine that you're working on the development team for a bank. Your teammate Isabella wrote the following BankAccount class. 
+
+```c#
+class BankAccount
```
Comment (memcmahon) [inline Mod1/Lessons/Week5/Scope.md:131]: I really like this bank example - its a good use case that is a real-world scenario 👍

File: Mod1/Lessons/Week5/Scope.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+## Fixing Scope Errors
+
+With your partner, take 15 minutes to walk through this training that introduces a common pattern for fixing scope errors, then discuss the questions below.
+
+https://learn.microsoft.com/en-us/training/modules/csharp-code-blocks/2-exercise-variable-scope
+
+> In your own words, why did the code in step 2 have an error? </br> In your own words, how did you fix the scope error created in step 2? </br> What would happen if scope didn't exist? What problems might this cause?
+
+
+<!-- OPEN QUESTION: Any reason for students to know the difference between method, class, block scope. Seems to all act the same. Curly brackets cause a new code block with a new scope. -->
```
Comment (memcmahon) [inline Mod1/Lessons/Week5/Scope.md:110]: I don't think we need to go into any more detail on class v method v scope.

File: Mod1/Lessons/Week5/Scope.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+Do I need code outside of the class to have access to this variable or method? (if yes, make it public)
+
+Is this an internal detail that should not be accessed from outside of the class? (if yes, make it private)
+
+<!-- Not sure we should talk about how hiding implementation details is an OOP pattern, encapsalation. Or leave that for a later day and reference public vs private as an example. -->
+
+You *could* make all of your variables and methods public and the code would still work, but as we saw, that can lead to unexpected uses of the class and headaches down the line.
+
+C# provides other types of access modifiers for more specific situations such as protected and internal. If you find yourself wanting a level of access between public and private, you can learn more in the documentation [here](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/access-modifiers). 
+
```
Comment (memcmahon) [inline Mod1/Lessons/Week5/Scope.md:225]: I think this overview of public/private is great - it focuses on the definitions without going into too much discussion of nuance.  I think Mod1 students will be able to pick up this concept and build on it.

File: Mod1/Lessons/Week5/Scope.md (hunk 2 above)
Comment (memcmahon) [reply Mod1/Lessons/Week5/Scope.md:59]: Love that you're thinking about this!

One thing that might help is to use different outlining for each box - not sure how easy that might be, but we could do something like outer box dashed, next box dotted, next box solid, next box dashed. just another visual differentiator between each box.

File: Mod1/Lessons/Week5/Scope.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+Maybe you've seen something like this where you already defined a variable but are still getting an error saying your variable doesn't exist.
+
+Example 1:
+![Scope Example 1](/Mod1/Images/Week5/ScopeExample1.png)
+
+Or maybe you've tried to use the variable i multiple times in your program and found that you need to create a new variable each time.
+
+Example 2:
+![Scope Example 2](/Mod1/Images/Week5/ScopeExample2.png)
```
Comment (memcmahon) [inline Mod1/Lessons/Week5/Scope.md:27]: This image is one I might change from an accessibility standpoint.  Orange and red are very close to each other on the color wheel and might be more difficult to differentiate than something like orange and green.

**Review by memcmahon — COMMENTED** [review]
This is great!

Just a couple of things you _could_ change, if you want to:
* adding a warmup
* working on the accessibility of images

File: Mod1/Lessons/Week5/Scope.md (hunk 1 above)
Comment (zoefarrell) [reply Mod1/Lessons/Week5/Scope.md:6]: Ooo I like this distinction. I agree, we want everyone to understand the different and start thinking about the uses cases but it almost certainly will take a while to make sense.

File: Mod1/Lessons/Week5/Scope.md (hunk 2 above)
Comment (zoefarrell) [reply Mod1/Lessons/Week5/Scope.md:59]: I'll give that a go! It's easy in excalidraw to use different borders. I'll take a look to see how visable the different borders are.
